cql/src/main/java/org/apache/cassandra/stargate/transport/internal/Server.java

```diff
@@ -622,41 +615,39 @@ private void send(Event event) {
       server.connectionTracker.send(event);
     }
 
-    // TODO(mpenick): Filter out storage nodes for status/topology events, clients don't need to
-    // "see" non-stargate node events
-
     @Override
-    public void onJoinCluster(InetAddressAndPort endpoint) {
-      if (server.persistence.isRpcReady(endpoint)) endpointsPendingJoinedNotification.add(endpoint);
-      else
-        onTopologyChange(
-            endpoint, Event.TopologyChange.newNode(server.persistence.getNativeAddress(endpoint)));
+    public void onJoinCluster(InetAddress endpoint, int port) {
+      InetAddressAndPort endpointWithPort = addPort(endpoint, port);
+      onTopologyChange(endpointWithPort, Event.TopologyChange.newNode(endpointWithPort));
     }
 
     @Override
-    public void onLeaveCluster(InetAddressAndPort endpoint) {
-      onTopologyChange(
-          endpoint,
-          Event.TopologyChange.removedNode(server.persistence.getNativeAddress(endpoint)));
+    public void onLeaveCluster(InetAddress endpoint, int port) {
+      InetAddressAndPort endpointWithPort = addPort(endpoint, port);
+      onTopologyChange(endpointWithPort, Event.TopologyChange.removedNode(endpointWithPort));
     }
 
     @Override
-    public void onMove(InetAddressAndPort endpoint) {
-      onTopologyChange(
-          endpoint, Event.TopologyChange.movedNode(server.persistence.getNativeAddress(endpoint)));
+    public void onMove(InetAddress endpoint, int port) {
+      InetAddressAndPort endpointWithPort = addPort(endpoint, port);
+      onTopologyChange(endpointWithPort, Event.TopologyChange.movedNode(endpointWithPort));
     }
 
     @Override
-    public void onUp(InetAddressAndPort endpoint) {
-      if (endpointsPendingJoinedNotification.remove(endpoint)) onJoinCluster(endpoint);
+    public void onDown(InetAddress endpoint, int port) {
+      InetAddressAndPort endpointWithPort = addPort(endpoint, port);
+      onStatusChange(endpointWithPort, Event.StatusChange.nodeDown(endpointWithPort));
+    }
 
-      onStatusChange(
-          endpoint, Event.StatusChange.nodeUp(server.persistence.getNativeAddress(endpoint)));
+    @Override
+    public void onUp(InetAddress endpoint, int port) {
+      InetAddressAndPort endpointWithPort = addPort(endpoint, port);
+      onStatusChange(endpointWithPort, Event.StatusChange.nodeUp(endpointWithPort));
     }
 
-    public void onDown(InetAddressAndPort endpoint) {
-      onStatusChange(
-          endpoint, Event.StatusChange.nodeDown(server.persistence.getNativeAddress(endpoint)));
+    private InetAddressAndPort addPort(InetAddress endpoint, int port) {
+      return InetAddressAndPort.getByAddressOverrideDefaults(
+          endpoint, port == NO_PORT ? server.socket.getPort() : port);
```

Comment:
so it basically takes the default port if there is no explicit port, right?
nit: maybe we could extract `port == NO_PORT ? server.socket.getPort() : port` to a separate method: `getPortOrDefault`.

Reply:
Good suggestion.